tools/clang/lib/SPIRV/SPIRVEmitter.cpp

```diff
@@ -1288,6 +1288,65 @@ uint32_t SPIRVEmitter::doConditionalOperator(const ConditionalOperator *expr) {
   return theBuilder.createSelect(type, condition, trueBranch, falseBranch);
 }
 
+uint32_t
+SPIRVEmitter::processByteAddressBufferLoad(const CXXMemberCallExpr *expr,
+                                           uint32_t numWords) {
+  const auto object = expr->getImplicitObjectArgument();
+  const auto type = object->getType();
+  assert(typeTranslator.isRWByteAddressBuffer(type) ||
+         typeTranslator.isByteAddressBuffer(type));
+  assert(numWords == 1 || numWords == 2 || numWords == 3 || numWords == 4);
+
+  if (expr->getNumArgs() == 1) {
+    const Expr *addressExpr = expr->getArg(0);
+    const uint32_t byteAddress = doExpr(addressExpr);
+    const uint32_t addressTypeId =
+        typeTranslator.translateType(addressExpr->getType());
+
+    // Do a OpShiftRightLogical by 2 (divide by 4 to get aligned memory
+    // access). The AST always casts the address to unsinged integer, so shift
+    // by unsinged integer 2.
+    const uint32_t constUint2 = theBuilder.getConstantUint32(2);
+    const uint32_t address = theBuilder.createShiftRightLogical(
+        addressTypeId, byteAddress, constUint2);
+
+    // Perform access chain into the (RW)ByteAddressBuffer.
+    // First index must be zero (member 0 of the struct is a
+    // runtimeArray). The second index passed to OpAccessChain should be
+    // the address.
+    const uint32_t uintTypeId = theBuilder.getUint32Type();
+    const uint32_t ptrType = theBuilder.getPointerType(
+        uintTypeId, declIdMapper.resolveStorageClass(object));
+    const uint32_t constUint0 = theBuilder.getConstantUint32(0);
+    const uint32_t loadPtr = theBuilder.createAccessChain(
+        ptrType, doExpr(object), {constUint0, address});
+    const uint32_t firstWord = theBuilder.createLoad(uintTypeId, loadPtr);
+    if (numWords == 1) {
+      return firstWord;
+    } else {
+      // Load word 2, 3, and 4 where necessary. Use OpCompositeConstruct to
+      // return a vector result.
+      llvm::SmallVector<uint32_t, 4> values;
+      values.push_back(firstWord);
+      for (uint32_t wordCounter = 2; wordCounter <= numWords; ++wordCounter) {
+        const uint32_t offset = theBuilder.getConstantUint32(wordCounter - 1);
+        const uint32_t newAddress = theBuilder.createBinaryOp(
+            spv::Op::OpIAdd, addressTypeId, address, offset);
+        const uint32_t loadPtr = theBuilder.createAccessChain(
```

Comment:
Hmm, variable shadowing. I would typically avoid this by just reusing `loadPtr` previously defined. I suspect some compilers will warn about this? (I mean future Linux support.)